1) Add PhasedFSimParameters container that describes cirq.PhasedFSimGate angles

mrwojtek opened the pull request:
This change adds PhasedFSimParameters that is a container for all the angles of cirq.PhasedFSimGate.

It is used for calibration purposes and stores only a numerical values for the angles, Since some calibrations estimate only a subset of the parameters, all of them are optional.

dstrain115 commented on cirq/google/calibration/phased_fsim.py:
```diff
@@ -0,0 +1,78 @@
+
+
+@json_serializable_dataclass(frozen=True)
+class PhasedFSimParameters:
```
Add doc string.

dstrain115 commented on cirq/protocols/json_serialization.py:
```diff
@@ -140,6 +140,7 @@ def two_qubit_matrix_gate(matrix):
         'PhaseFlipChannel': cirq.PhaseFlipChannel,
         'PhaseGradientGate': cirq.PhaseGradientGate,
         'PhasedFSimGate': cirq.PhasedFSimGate,
+        'PhasedFSimParameters': cirq.google.PhasedFSimParameters,
```
If you want, you can add this to the NOT_YET_SERIALIZED list, and I can add the json serialization if you want?

mrwojtek replied:
There is no problem with json serialization for this dataclass as far as I know (I'll double-check this). The problems are in PR 2)

mrwojtek commented:
Update:
1. Rename from PhsaedFSimParameters to PhasedFSimCharacterization. This way it becomes calibration internal dataclass which most likely should be kept here. The main difference between here and the rest of cirq is that symbols are meaningless and unitary angles might be optional, which is something meaningless for the gate notion.
2. Added complete docstrings.
3. Added tests.
4. Requires styling/formatting changes.

mrwojtek commented:
I have tested the JSON serialization/deserialization, it works. Would be nice to have a tests for that in the future.

dstrain115 commented on cirq/google/calibration/phased_fsim.py:
```diff
@@ -0,0 +1,106 @@
+
+    This class stores five unitary parameters (θ, ζ, χ, γ and φ) that describe the
+    cirq.PhasedFSimGate which is the most general particle conserving two-qubit gate. The unitary
+    of the underlying tate is:
```
tate to gate

dstrain115 commented on cirq/google/calibration/phased_fsim.py:
```diff
@@ -0,0 +1,106 @@
+    phi: Optional[float] = None
+
+    def asdict(self) -> Dict[str, float]:
+        """Converts parameters to a dictionary that maps angles name to values.
```
nit: I would probably change to "angle names"

dstrain115 commented on cirq/google/calibration/phased_fsim.py:
```diff
@@ -0,0 +1,106 @@
+        return dataclasses.asdict(self)
+
+    def all_none(self) -> bool:
+        """Checks if all the angles are None."""
```
Maybe "Returns True if all the angles are None." would be more precise.

Same below for any_none

dstrain115 reviewed (approved):
I am surprised that you don't have to add a json and repr file to json_test_data for this.  I guess @json_serialized_dataclass takes care of it for you.

Some minor nits, but otherwise, looks fine to me.

mrwojtek commented:
You are right with JSON serialization. It was working but the Cirq serialization tests were failing since I didn't provide the necessary golden files apparently. I removed the serialization completely from this PR, this is something to be done in a follow-up.

mpharrigan commented:
Is there a way to see a clean diff of this PR? It's pulling in all the changes to Cirq `master` as well

mpharrigan commented:
This PR is https://github.com/quantumlib/Cirq/pull/3646, right?

mrwojtek commented:
Yes
